internal/fs/vss_windows.go

```diff
@@ -0,0 +1,870 @@
+	var oleIUnknown *ole.IUnknown
+	result, _, _ := syscall.Syscall(vss.getVTable().gatherWriterMetadata, 2,
+		uintptr(unsafe.Pointer(vss)), uintptr(unsafe.Pointer(&oleIUnknown)), 0)
+	return vss.convertToVSSAsync(result, oleIUnknown)
+}
+
+// IsVolumeSupported calls the equivalent VSS api.
+func (vss *IVssBackupComponents) IsVolumeSupported(volumeName string) (HRESULT, bool) {
+	volumeNamePointer, err := syscall.UTF16PtrFromString(volumeName)
+	if err != nil {
+		panic(err)
+	}
+
+	isSupported := false
+	var result uintptr = 0
+
+	if runtime.GOARCH == "386" {
+		result, _, _ = syscall.Syscall9(vss.getVTable().isVolumeSupported, 7,
+			uintptr(unsafe.Pointer(vss)), uintptr(ole.IID_NULL.Data1), uintptr(ole.IID_NULL.Data2),
+			uintptr(ole.IID_NULL.Data3), uintptr(ole.IID_NULL.Data4[0]),
+			uintptr(unsafe.Pointer(volumeNamePointer)), uintptr(unsafe.Pointer(&isSupported)), 0, 0)
+	} else {
+		result, _, _ = syscall.Syscall6(vss.getVTable().isVolumeSupported, 4,
+			uintptr(unsafe.Pointer(vss)), uintptr(unsafe.Pointer(ole.IID_NULL)),
+			uintptr(unsafe.Pointer(volumeNamePointer)), uintptr(unsafe.Pointer(&isSupported)), 0, 0)
+	}
+	return HRESULT(result), isSupported
+}
+
+// StartSnapshotSet calls the equivalent VSS api.
+func (vss *IVssBackupComponents) StartSnapshotSet() (HRESULT, ole.GUID) {
+	var snapshotSetId ole.GUID
+	result, _, _ := syscall.Syscall(vss.getVTable().startSnapshotSet, 2,
+		uintptr(unsafe.Pointer(vss)), uintptr(unsafe.Pointer(&snapshotSetId)), 0,
+	)
+	return HRESULT(result), snapshotSetId
+}
+
+// AddToSnapshotSet calls the equivalent VSS api.
+func (vss *IVssBackupComponents) AddToSnapshotSet(volumeName string, idSnapshot *ole.GUID) HRESULT {
+	volumeNamePointer, err := syscall.UTF16PtrFromString(volumeName)
+	if err != nil {
+		panic(err)
+	}
+
+	var result uintptr = 0
+
+	if runtime.GOARCH == "386" {
+		result, _, _ = syscall.Syscall9(vss.getVTable().addToSnapshotSet, 7,
+			uintptr(unsafe.Pointer(vss)), uintptr(unsafe.Pointer(volumeNamePointer)),
+			uintptr(ole.IID_NULL.Data1), uintptr(ole.IID_NULL.Data2), uintptr(ole.IID_NULL.Data3),
+			uintptr(ole.IID_NULL.Data4[0]), uintptr(unsafe.Pointer(idSnapshot)), 0, 0)
+	} else {
+		result, _, _ = syscall.Syscall6(vss.getVTable().addToSnapshotSet, 4,
+			uintptr(unsafe.Pointer(vss)), uintptr(unsafe.Pointer(volumeNamePointer)),
+			uintptr(unsafe.Pointer(ole.IID_NULL)), uintptr(unsafe.Pointer(idSnapshot)), 0, 0)
+	}
+
+	return HRESULT(result)
+}
+
+// PrepareForBackup calls the equivalent VSS api.
+func (vss *IVssBackupComponents) PrepareForBackup() (HRESULT, *IVSSAsync) {
+	var oleIUnknown *ole.IUnknown
+	result, _, _ := syscall.Syscall(vss.getVTable().prepareForBackup, 2,
+		uintptr(unsafe.Pointer(vss)), uintptr(unsafe.Pointer(&oleIUnknown)), 0)
+
+	return vss.convertToVSSAsync(result, oleIUnknown)
+}
+
+// convertToVSSAsync looks up IVSSAsync interface if given result
+// is a success.
+func (vss *IVssBackupComponents) convertToVSSAsync(
```

Comment:
The first call of this method is by `GatherWriterMetadata`.

Reply:
Can you please explain what should be changed here?

Reply:
I would have expected that `convertToVSSAsync` is placed right below the `GatherWriterMetadata` function, as that one is the first to call `convertToVSSAsync` and not `PrepareForBackup`.

Reply:
Fixed in https://github.com/restic/restic/pull/2274/commits/602ce26f035dd15856e1464ba616ce561d4a8380